File: services/chess_service/src/main/java/net/chess_platform/chess_service/chess/move/EnPassantMove.java
package net.chess_platform.chess_service.chess.move;

import net.chess_platform.chess_service.chess.Chessboard;
import net.chess_platform.chess_service.chess.piece.Piece;
import net.chess_platform.chess_service.chess.piece.Piece.Color;

public class EnPassantMove extends AbstractMove {

    private Piece capturedPawnInstance;

    private Position capturedPawnPosition;

    public EnPassantMove(Piece.Type piece, Color color, Position from, Position to) {
        super(piece, color, from, to, Type.EN_PASSANT);
    }

    @Override
    public void execute(Chessboard board) {
        super.execute(board);
        var movedPiece = getMovedPieceInstance();
        var from = getFrom();
        var to = getTo();

        capturedPawnPosition = new Position(to.row() - Chessboard.getPawnDirection(movedPiece.getColor()), to.col());
        capturedPawnInstance = board.getPiece(capturedPawnPosition);

        board.setPiece(movedPiece, to);
        board.setPiece(null, from);
        board.setPiece(null, capturedPawnPosition);
    }

    @Override
    public void undo(Chessboard board) {
        var movedPiece = getMovedPieceInstance();
        var from = getFrom();
        var to = getTo();

        board.setPiece(movedPiece, from);
        board.setPiece(null, to);
        board.setPiece(capturedPawnInstance, capturedPawnPosition);

        capturedPawnPosition = null;
        capturedPawnInstance = null;

        super.undo(board);
    }

}

File: app/src/chess/moves/EnPassantMove.ts
import { toAlgebraicNotation, toCodePoint } from "../chess-view";
import { Chessboard } from "../Chessboard";
import { Color } from "../pieces/AbstractPiece";
import { Piece, PieceType } from "../pieces/Piece";
import { AbstractMove } from "./AbstractMove";
import { CheckStatus, Position } from "./Move";

export class EnPassantMove extends AbstractMove {

    private capturedPawnInstance?: Piece;

    constructor(piece: PieceType | null, color: Color, from: Position, to: Position, checkStatus?: CheckStatus, timestamp?: number) {
        super(piece, color, from, to, "EN_PASSANT", checkStatus, timestamp);
    }

    override execute(board: Chessboard): void {
        super.execute(board);

        const movedPiece = this.getMovedPieceInstance()!;
        const from = this.getFrom();
        const to = this.getTo();

        const capturedPos = this.getCapturedPawnPosition();
        this.capturedPawnInstance = board.getPiece(capturedPos.row, capturedPos.col)!;

        board.setPiece(to.row, to.col, movedPiece);
        board.setPiece(from.row, from.col, null);
        board.setPiece(capturedPos.row, capturedPos.col, null);
    }

    override undo(board: Chessboard): void {
        const movedPiece = this.getMovedPieceInstance()!;
        const capturedPos = this.getCapturedPawnPosition();
        const from = this.getFrom();
        const to = this.getTo();

        board.setPiece(from.row, from.col, movedPiece);
        board.setPiece(to.row, to.col, null);
        board.setPiece(capturedPos.row, capturedPos.col, this.capturedPawnInstance!);

        this.capturedPawnInstance = undefined;

        super.undo(board);
    }

    getCapturedPawnPosition(): Position {
        const to = this.getTo();
        return { row: to.row - Chessboard.getPawnDirection(this.getColor()), col: to.col };
    }

    toAlgebraicNotationImpl(): string {
        return `${toCodePoint(this.getColor() + "-" + this.getPiece())}${toAlgebraicNotation(this.getFrom())}x${toAlgebraicNotation(this.getTo())}e.p.`;
    }

}